### PR TITLE
Allow for unlimited bitrate

### LINKED ISSUE #1015: High Bitrate H264 Videos Transcode, Reason: The Video's Bitrate Not Supported
I've discovered that some of my movies that are encoded in h264 and have a video bitrate that exceeds 10000kbps will result in transcoding.  The transcoding also knocks the 5.1 passthrough track down to AAC 2.0.  I can't say specifically which release these files last worked on, but they 100% used to direct play without issue. I'm doing a bit more investigating and will update with any further findings. 

DTS

### PR DESCRIPTION
**Pull request by jimdogx**
Just giving the people what they want.  Some users swear that without capping their bitrate, their devices can direct stream videos that have a higher bitrate than what Roku says is allowed.

**Changes**
~~New setting that will not send the max bitrate to the Jellyfin server.  With the default being OFF.~~
* New setting to enable / disable bitrate limiting.  With the default being ON.
* New setting to enter a max bit rate when limiting is enabled.  If set to 0, the specs in Roku's docs will be used.

**Issues**
Users that don't understand this setting my turn it on and have issues with playback.

**Fixes**
Fixes #1015

### CODE REVIEW

File: source/utils/deviceCapabilities.brs [hunk 1]
```diff
@@ -382,3 +358,48 @@ function GetDirectPlayProfiles() as object
...
 
 end function
+
+function GetBitRateLimit(codec as string)
+
+    ' Some repeated values (e.g. same "40mbps" for several codecs)
+    ' but this makes it easy to update in the future if the bitrates start to deviate.
+    if get_user_setting("playback.bitrate.unlimited") = "true"
+        return {}
+    else
```
Comment (1hitsong) [inline source/utils/deviceCapabilities.brs]: else isn't needed. If we fall into the main condition, we're returned.

**Review by 1hitsong — CHANGES_REQUESTED** [review]

File: source/utils/deviceCapabilities.brs (hunk 1 above)
Comment (jimdogx) [reply source/utils/deviceCapabilities.brs]: This is why I shouldn't code at 5 am before I've had my coffee or breakfast 😄

File: settings/settings.json [hunk 2]
```diff
@@ -63,6 +63,13 @@
                 "settingName": "playback.subs.onlytext",
                 "type": "bool",
                 "default": "false"
+            },
+            {
+                "title": "Turn off bitrate limiting",
```
Comment (1hitsong) [inline settings/settings.json]:
```suggestion
                "title": "Bitrate Limiting",
```

File: settings/settings.json [hunk 3]
```diff
@@ -63,6 +63,13 @@
...
+                "description": "This will ignore Roku's official max bitrate specifications and attempt to stream without enforcing a max bitrate.",
```
Comment (1hitsong) [inline settings/settings.json]:
```suggestion
                "description": "Enable or disable limiting video bitrates based on Roku's specifications.",
```

File: settings/settings.json [hunk 4]
```diff
@@ -63,6 +63,13 @@
...
+                "settingName": "playback.bitrate.unlimited",
+                "type": "bool",
+                "default": "false"
```
Comment (1hitsong) [inline settings/settings.json]:
```suggestion
                "default": "true"
```

File: settings/settings.json [hunk 5]
```diff
@@ -63,6 +63,13 @@
                 "settingName": "playback.subs.onlytext",
                 "type": "bool",
                 "default": "false"
+            },
+            {
+                "title": "Turn off bitrate limiting",
+                "description": "This will ignore Roku's official max bitrate specifications and attempt to stream without enforcing a max bitrate.",
+                "settingName": "playback.bitrate.unlimited",
```
Comment (1hitsong) [inline settings/settings.json]:
```suggestion
                "settingName": "playback.bitrate.maxlimited",
```

File: source/utils/deviceCapabilities.brs [hunk 6]
```diff
@@ -382,3 +358,49 @@ function GetDirectPlayProfiles() as object
     ]
 
 end function
+
+function GetBitRateLimit(codec as string)
+
+    ' If the user has requested to not cap their bitrate (regardless of if that works or not)...
+    if get_user_setting("playback.bitrate.unlimited") = "true"
```
Comment (1hitsong) [inline source/utils/deviceCapabilities.brs]:
```suggestion
    if not get_user_setting("playback.bitrate.maxlimited") = "true"
```

**Review by 1hitsong — CHANGES_REQUESTED** [review]
I reworded the setting a bit. The translations will need to be updated once finalized.

File: source/utils/deviceCapabilities.brs (hunk 6 above)
Comment (jimdogx) [reply source/utils/deviceCapabilities.brs]: I went ahead and adjusted the if block now that it's "maxlimited" instead of "unlimited" 👍

**Comment by NeuroDawg** [comment]
Thanks. I'm one of those folks who's Roku 4 would play h264 files with bitrates higher than Roku publishes. Ideally, however, would be the choice to set a bitrate limit. Without a limit, high quality 4K HEVC movies with bitrates of 60-90mbps would fail to play, while 1080p h264 files with bitrates of 40mpbs would (at least for me).

**Comment by candry7731** [comment]
Should we add an alert to change the setting if the video fails/errors because of the bitrate setting?

**Comment by crobibero** [comment]
Just an fyi- if the bitrate limit isnt set the server will default to 80Mbit

**Comment by jimdogx** [comment]
> Should we add an alert to change the setting if the video fails/errors because of the bitrate setting?

The default will be to limit it per Roku's specs.  I think most users won't be changing this setting.  For those that do, they most likely know what they're getting themselves in to 😄

**Comment by sevenrats** [comment]
can we please change "Bitrate Limiting" to "Limit Bitrate"  this will better match our current style of using imperatives instead of infinitives.

File: source/utils/deviceCapabilities.brs [hunk 7]
```diff
@@ -265,17 +247,18 @@ function getDeviceProfile() as object
...
-                    "Property": "VideoBitrate",
-                    "Value": "40000000",
-                    IsRequired: true
-                }
+                GetBitRateLimit("VP9")
             ]
         })
     end if
 
+    print "DEBUG: "
```
Comment (1hitsong) [inline source/utils/deviceCapabilities.brs]: Debug code 255 - 261

**Review by 1hitsong — CHANGES_REQUESTED** [review]

File: locale/en_US/translations.ts [hunk 8]
```diff
@@ -1021,6 +1021,22 @@
...
+        <message>
+            <source>Enable or disable limiting video bitrates (based on the 'Playback Bitrate Limit' setting *or* Roku's specifications).</source>
+            <translation>Enable or disable limiting video bitrates (based on the 'Playback Bitrate Limit' setting *or* Roku's specifications).</translation>
+        </message>
+        <message>
+            <source>Playback Bitrate Limit</source>
+            <translation>Playback Bitrate Limit</translation>
+        </message>
+        <message>
+            <source>Max bitrate (MBPS) to use if 'Limit Bitrate' is enabled. If set to 0 limits will be per Roku's specifications.</source>
```
Comment (1hitsong) [inline locale/en_US/translations.ts]:
```suggestion
            <source>Max bitrate (MBPS) to use if 'Limit Bitrate' is enabled. Set to 0 to use Roku's specifications.</source>
```

File: locale/en_US/translations.ts [hunk 9]
```diff
@@ -1021,6 +1021,22 @@
...
+            <translation>Max bitrate (MBPS) to use if 'Limit Bitrate' is enabled. If set to 0 limits will be per Roku's specifications.</translation>
```
Comment (1hitsong) [inline locale/en_US/translations.ts]:
```suggestion
            <translation>Max bitrate (MBPS) to use if 'Limit Bitrate' is enabled. Set to 0 to use Roku's specifications.</translation>
```

File: locale/en_US/translations.ts [hunk 10]
```diff
@@ -1021,6 +1021,22 @@
             <source>If enabled, the star and community rating for episodes of a TV show will be removed. This is to prevent spoilers of an upcoming good/bad episode.</source>
             <translation>If enabled, the star and community rating for episodes of a TV show will be removed. This is to prevent spoilers of an upcoming good/bad episode.</translation>
         </message>
+        <message>
+            <source>Limit Bitrate</source>
+            <translation>Limit Bitrate</translation>
+        </message>
+        <message>
+            <source>Enable or disable limiting video bitrates (based on the 'Playback Bitrate Limit' setting *or* Roku's specifications).</source>
```
Comment (1hitsong) [inline locale/en_US/translations.ts]:
```suggestion
            <source>If enabled, playback bitrates will be limited based on the 'Playback Bitrate Limit' setting.</source>
```

File: locale/en_US/translations.ts [hunk 11]
```diff
@@ -1021,6 +1021,22 @@
...
+            <translation>Enable or disable limiting video bitrates (based on the 'Playback Bitrate Limit' setting *or* Roku's specifications).</translation>
```
Comment (1hitsong) [inline locale/en_US/translations.ts]:
```suggestion
            <translation>If enabled, playback bitrates will be limited based on the 'Playback Bitrate Limit' setting.</translation>
```

**Review by 1hitsong — CHANGES_REQUESTED** [review]

File: source/utils/deviceCapabilities.brs (hunk 7 above)
Comment (jimdogx) [reply source/utils/deviceCapabilities.brs]: @#$^@#ADF@%

File: settings/settings.json [hunk 12]
```diff
@@ -63,6 +63,20 @@
                 "settingName": "playback.subs.onlytext",
                 "type": "bool",
                 "default": "false"
+            },
+            {
+                "title": "Limit Bitrate",
```
Comment (sevenrats) [inline settings/settings.json]: would you please insert these settings in alphabetical order?  It goes General Menu, Menus, Settings.

Comment (jimdogx) [reply settings/settings.json]: I think @1hitsong just did this for me 👍

**Review by 1hitsong — APPROVED** [review]

**Pull request merged**